### PR TITLE
[Feature] Add mean pooling support for embedding pooler

### PR DESCRIPTION
**Pull request by smfirmin**
## Motivation

  This is PR 1 of an incremental `nomic-ai` BERT embedding onboarding.

  It adds the generic `MEAN` pooling primitive needed for Nomic-style embedding models and covers it with
  focused unit tests. The implementation also avoids an unnecessary device-to-host sync in the normal
  forward path by using `extend_seq_lens_cpu` when available.

  This PR is intentionally narrow and does not claim full `nomic-ai/nomic-embed-text-v1.5` support yet. It
  establishes the reusable pooling behavior that later PRs can wire into a Nomic-compatible BERT-with-RoPE
  model path and model-specific post-processing.

  ## Modifications

  - add `PoolingType.MEAN`
  - implement mean pooling in `Pooler.forward()`
  - prefer `forward_batch.extend_seq_lens_cpu` to avoid a hot-path sync
  - support both list-backed and tensor-backed `extend_seq_lens_cpu`
  - add focused unit tests for:
    - `LAST`
    - `CLS`
    - `MEAN`
    - normalization
    - dimension truncation
    - mixed-dimension list outputs
    - tensor-backed `extend_seq_lens_cpu`

  ## Validation

  Ran locally:

  ```bash
  python -m py_compile python/sglang/srt/layers/pooler.py test/registered/unit/layers/test_pooler.py
  PYTHONPATH=python ./.venv/bin/python -c 'from sglang._triton_stub import install; install(); import
  pytest, sys; sys.exit(pytest.main(["test/registered/unit/layers/test_pooler.py", "-q"]))'
```
  Result: 7 passed

  ## Checklist

  - [x] Format your code according to the Format code with pre-commit
    (https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
  - [x] Add unit tests according to the Run and add unit tests
    (https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
  - [ ] Update documentation according to Write documentations
    (https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
  - [ ] Provide accuracy and speed benchmark results according to Test the accuracy
    (https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and Benchmark the sp
    eed (https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
  - [x] Follow the SGLang code style guidance
    (https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

  ## Review and Merge Process

  1. Ping Merge Oncalls to start the process. See the PR Merge Process
     (https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
  2. Get approvals from CODEOWNERS (https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and
     other reviewers.
  3. Trigger CI tests with comments
     (https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact au
     thorized users to do so.
  4. After green CI and required approvals, ask Merge Oncalls or people with write permission to merge the
     PR.

  Common CI commands:

  - /tag-and-rerun-ci
  - /tag-run-ci-label
  - /rerun-failed-ci

  ## Follow-ups

  Planned follow-up PRs will add:

  - a Nomic-compatible BERT model path / architecture registration
  - wiring that selects MEAN pooling for that model
  - Nomic-specific embedding post-processing and request semantics where needed

### CODE REVIEW

File: python/sglang/srt/layers/pooler.py [hunk 1]
```diff
@@ -53,6 +54,20 @@ def forward(
...
             pooled_data = hidden_states[first_token_flat_indices]
+        elif self.pooling_type == PoolingType.MEAN:
+            pooled_data_lst = []
+            offset = 0
+            prompt_lens = getattr(forward_batch, "extend_seq_lens_cpu", None)
+            if prompt_lens is None:
+                prompt_lens = forward_batch.extend_seq_lens.tolist()
+            elif torch.is_tensor(prompt_lens):
+                prompt_lens = prompt_lens.tolist()
+            for prompt_len in prompt_lens:
```
Comment (alexnails) [inline python/sglang/srt/layers/pooler.py]:
```
elif self.pooling_type == PoolingType.MEAN:
    prompt_lens = getattr(forward_batch, "extend_seq_lens_cpu", forward_batch.extend_seq_lens)
    if torch.is_tensor(prompt_lens):
        prompt_lens = prompt_lens.tolist()
    pooled_data = torch.stack([
        seq.mean(dim=0) for seq in torch.split(hidden_states, prompt_lens)
    ])
```

e.g can we vectorize?

Comment (smfirmin) [reply python/sglang/srt/layers/pooler.py]: thank you for reviewing! Please see the changes I pushed and let me know what you think.

File: python/sglang/srt/layers/pooler.py [hunk 2]
```diff
@@ -53,6 +54,25 @@ def forward(
             first_token_flat_indices = torch.zeros_like(prompt_lens)
             first_token_flat_indices[1:] += torch.cumsum(prompt_lens, dim=0)[:-1]
             pooled_data = hidden_states[first_token_flat_indices]
+        elif self.pooling_type == PoolingType.MEAN:
```
Comment (alexnails) [inline python/sglang/srt/layers/pooler.py]: I think this looks correct, but is there reason for forcing fp32 / int64?

Comment (smfirmin) [reply python/sglang/srt/layers/pooler.py]: Yes -  Mean pooling is sensitive to accumulation error on bf16/fp16 inputs, so the implementation accumulates and divides in float32 to keep embedding quality stable. The int64 forcing is because _prompt_lens_ feeds _repeat_interleave_ and then _index_add__, which becomes indexing metadata, and long is the safest dtype for that.

Comment (alexnails) [reply python/sglang/srt/layers/pooler.py]: torch.repeat_interleave / int64 usage does not benefit here (this would come from things like forward batch, etc which is already in int32) so afaik we really do not keep anything in int64. The fp32 stuff makes sense. @smfirmin please remove the int64 stuff and I will approve and run CI

Comment (smfirmin) [reply python/sglang/srt/layers/pooler.py]: done - thank you!

**Comment by alexnails** [comment]
@smfirmin sorry for the ask, but could you fix the merge conflict and then we merge up to date with main? I will ping people to merge this into main as I know it has been sitting

File: python/sglang/srt/layers/pooler.py [hunk 3]
```diff
@@ -61,6 +62,20 @@ def pool_hidden_states(
         first_token_flat_indices = torch.zeros_like(prompt_lens)
         first_token_flat_indices[1:] += torch.cumsum(prompt_lens, dim=0)[:-1]
         return hidden_states[first_token_flat_indices]
+    elif pooling_type == PoolingType.MEAN:
+        prompt_lens = forward_batch.extend_seq_lens.to(device=hidden_states.device)
```
Comment (JustinTong0323) [inline python/sglang/srt/layers/pooler.py]: The PR description says "prefer `extend_seq_lens_cpu` to avoid a hot-path sync" and "support both list-backed and tensor-backed `extend_seq_lens_cpu`", but this branch reads `extend_seq_lens` (the GPU tensor) and then calls `torch.repeat_interleave(arange, prompt_lens)` with no `output_size=`. On CUDA that forces a D→H sync to compute the output shape — exactly the sync the description says it avoids.

Either switch to `extend_seq_lens_cpu` (handling list vs tensor) and pass `output_size=int(prompt_lens_cpu.sum())`, or drop that claim from the description.

Nit while you're here: `.to(device=hidden_states.device)` is a no-op in the real pipeline (`extend_seq_lens` is already on device) and inconsistent with the LAST/CLS branches above — it goes away if you switch to `extend_seq_lens_cpu`.

File: python/sglang/srt/layers/pooler.py [hunk 4]
```diff
@@ -61,6 +62,20 @@ def pool_hidden_states(
...
+            torch.arange(prompt_lens.shape[0], device=hidden_states.device),
+            prompt_lens,
+        )
+        pooled_data = torch.zeros(
+            (prompt_lens.shape[0], hidden_states.shape[-1]),
+            dtype=torch.float32,
+            device=hidden_states.device,
+        )
+        pooled_data.index_add_(0, batch_indices, hidden_states.to(dtype=torch.float32))
+        pooled_data /= prompt_lens.to(dtype=torch.float32).unsqueeze(-1)
```
Comment (JustinTong0323) [inline python/sglang/srt/layers/pooler.py:94]: Worth one line in the `pool_hidden_states` docstring: MEAN returns float32 regardless of input dtype, while LAST/CLS preserve the input dtype. Easy to surprise a caller otherwise.

File: test/registered/unit/layers/test_pooler.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+from types import SimpleNamespace
+
+from sglang.test.ci.ci_register import register_cpu_ci
+
+register_cpu_ci(est_time=5, suite="stage-a-test-cpu")
+
+import torch
+
+
+def _load_pooler_module():
```
Comment (JustinTong0323) [inline test/registered/unit/layers/test_pooler.py]: This ~45-line `_load_pooler_module` importlib / `sys.modules` shim isn't needed. `register_cpu_ci` already pulls in `sglang`, and the sibling `test_pooler_score_and_pool.py` in this same directory runs on the same CPU suite with a plain `from sglang.srt.layers.pooler import ...`. Drop the shim and use a normal import — halves the file and makes `Pooler`/`PoolingType` greppable.

File: test/registered/unit/layers/test_pooler.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+        extend_seq_lens_cpu=(
+            list(extend_seq_lens)
+            if extend_seq_lens_cpu is None
+            else extend_seq_lens_cpu
+        ),
+        dimensions=dimensions,
+    )
+
+
+class TestPooler(unittest.TestCase):
```
Comment (JustinTong0323) [inline test/registered/unit/layers/test_pooler.py]: Subclass `CustomTestCase` from `sglang.test.test_utils`, not bare `unittest.TestCase` — project convention, and `test_pooler_score_and_pool.py` in the same directory already follows it.

File: test/registered/unit/layers/test_pooler.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+                [
+                    [0.6, 0.8],
+                    [0.0, 1.0],
+                ]
+            ),
+            atol=1e-6,
+            rtol=1e-6,
+        )
+
+    def test_mean_pooling_supports_tensor_extend_seq_lens_cpu(self):
```
Comment (JustinTong0323) [inline test/registered/unit/layers/test_pooler.py:123]: This test is vacuous: the MEAN branch in `pool_hidden_states` never reads `extend_seq_lens_cpu` — only `extend_seq_lens`. Set `extend_seq_lens_cpu` to `None`, a string, or a dict here and the test still passes. The name claims a contract the code doesn't provide.

Delete it, or make it real once the sync-avoidance concern on `pooler.py` is fixed.

**Review by JustinTong0323 — COMMENTED** [review]
A few things before this lands — see inline.

**Comment by JustinTong0323** [comment]
Hi Seems you need to fix the lint CI

**Comment by JustinTong0323** [comment]
Hi Seems you need to fix the lint CI

**Comment by smfirmin** [comment]
Hi @JustinTong0323 I think I fixed the lint CI if you can take another look when you get a chance

**Comment by hnyls2002** [comment]
Thanks @smfirmin! MEAN pooling landed independently - `PoolingType.MEAN` is defined at `python/sglang/srt/layers/pooler.py:23` and `pool_hidden_states` implements it at `pooler.py:64-71` (cumsum-based rather than index_add_, same result), with coverage in `test/registered/unit/layers/test_pooler_score_and_pool.py`. Please reopen if I've missed something.
